File: 프로그래머스/1/160586. 대충 만든 자판/대충 만든 자판.cs
using System;
using System.Collections.Generic;

public class Solution 
{
    public int[] solution(string[] keymap, string[] targets) 
    {
        Dictionary<char, int> minPress = new Dictionary<char, int>();
        
        foreach(string key in keymap)
        {
            for (int i = 0; i < key.Length; i++)
            {
                char c = key[i];
                int press = i + 1;  // i번째 위치니까 (i+1)번 눌러야 함
                
                if (!minPress.ContainsKey(c) || minPress[c] > press)
                {
                    minPress[c] = press;  // 더 적은 횟수로 갱신
                }
            }
        }
        
        int[] result = new int[targets.Length];
        
        for(int i=0; i<targets.Length; i++)
        {
            int total = 0;
            bool possible = true;
            foreach(char c in targets[i])
            {
                if(!minPress.ContainsKey(c))
                {
                    possible = false;
                    break;
                }
                total += minPress[c];
            }
            result[i] = possible ? total : -1;
        }
        return result;
    }
}

File: 프로그래머스/1/160586. 대충 만든 자판/대충 만든 자판.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

vector<int> solution(vector<string> keymap, vector<string> targets) 
{
    map<char, int> touchmap;
    for(int i=0; i<keymap.size(); i++)
    {
        for(int j=0; j<keymap[i].size(); j++)
        {
            char c = keymap[i][j];
            
            if(touchmap.find(c) != touchmap.end())
            {
                touchmap[c] = min(touchmap[c], j + 1);
            }
            else
            {
                touchmap[c] = j + 1;
            }
        }
    }
    
    vector<int> result;
    for(string &target : targets)
    {
        int touchCount = 0;
        for(char &ch : target)
        {
            if(touchmap.find(ch) != touchmap.end())
            {
                touchCount += touchmap[ch];
            }
            else
            {
                touchCount = -1;
                break;
            }
        }
        result.push_back(touchCount);
    }
    return result;
}